File: app/src/main/java/com/ritesh/codeforcesportal/adapter/ContestAdapter.java
package com.ritesh.codeforcesportal.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.chip.Chip;
import com.ritesh.codeforcesportal.R;
import com.ritesh.codeforcesportal.model.Contest;
import com.ritesh.codeforcesportal.utils.Utils;

import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class ContestAdapter extends RecyclerView.Adapter<ContestAdapter.MyViewHolder>{

    private List<Contest> contestsList;
    private boolean isLimit;
    private final int LIMIT = 3;
    private final Context context;

    public ContestAdapter(Context context, boolean isLimit) {
        this.isLimit = isLimit;
        this.context = context;
    }

    public void updateList(List<Contest> list) {
        this.contestsList = list;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.contest_list_item,parent,false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        Contest contest = contestsList.get(position);
        holder.contestName.setText(contest.getName());
        holder.contestType.setText(contest.getType());
        holder.contestLength.setText(Utils.formatTimeInSeconds(contest.getDurationSeconds()));
        holder.contestStartTime.setText(Utils.formatContestStartTime(contest.getStartTimeSeconds()));
        holder.moreInfo.setOnClickListener(view -> {
            context.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://codeforces.com/contests/" + contest.getId())));
        });
    }

    @Override
    public int getItemCount() {
        if(contestsList == null) return 0;
        if(isLimit)
            return Math.min(LIMIT,contestsList.size());
        return contestsList.size();
    }

    public static class MyViewHolder extends RecyclerView.ViewHolder {
        TextView contestName, contestLength, contestStartTime;
        Chip contestType;
        CircleImageView moreInfo;
        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            contestName = itemView.findViewById(R.id.contest_name);
            contestLength = itemView.findViewById(R.id.contest_length);
            contestStartTime = itemView.findViewById(R.id.contest_start_time);
            contestType = itemView.findViewById(R.id.contest_type);
            moreInfo = itemView.findViewById(R.id.more_info);
        }
    }
}

File: app/src/main/java/com/ritesh/codeforcesvisualizer/ui/contest/ContestAdapter.kt
package com.ritesh.codeforcesvisualizer.ui.contest

import android.content.Intent
import android.net.Uri
import android.provider.CalendarContract
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.ritesh.codeforcesvisualizer.data.remote.CodeforcesApi
import com.ritesh.codeforcesvisualizer.databinding.ContestItemBinding
import com.ritesh.codeforcesvisualizer.model.Contest
import com.ritesh.codeforcesvisualizer.util.Utils.toFormattedTime
import java.text.SimpleDateFormat
import java.util.*

class ContestAdapter: ListAdapter<Contest,ContestAdapter.ContestViewHolder>(ContestItemDiffCallback()) {



    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContestViewHolder {
        val binding: ContestItemBinding = ContestItemBinding.inflate(LayoutInflater.from(parent.context))
        val params = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        binding.root.layoutParams = params
        return ContestViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ContestViewHolder, position: Int) {
        holder.onBind(getItem(position))
    }

    class ContestViewHolder(val binding: ContestItemBinding): RecyclerView.ViewHolder(binding.root) {
        fun onBind(contest: Contest) {
            binding.contestName.text = contest.name
            binding.contestStatus.text = contest.phase
            binding.contestStartTime.text = toFormattedTime(contest.startTime)

            binding.root.setOnClickListener {
                if (binding.moreOptions.isVisible)
                    binding.moreOptions.visibility = View.GONE
                else
                    binding.moreOptions.visibility = View.VISIBLE
            }

            binding.register.setOnClickListener {
                val url = "${CodeforcesApi.BASE_URL}/contestRegistration/${contest.id}"
                val intent = Intent().apply {
                    action = Intent.ACTION_VIEW
                    data = Uri.parse(url)
                }
                binding.root.context.startActivity(intent)
            }

            binding.reminder.setOnClickListener {

                val intent = Intent(Intent.ACTION_INSERT)
                    .setData(CalendarContract.Events.CONTENT_URI)
                    .putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME,contest.startTime)
                    .putExtra(CalendarContract.EXTRA_EVENT_END_TIME, contest.startTime + contest.duration)
                    .putExtra(CalendarContract.Events.TITLE, contest.name)
                binding.root.context.startActivity(intent)
            }
        }

    }

    class ContestItemDiffCallback: DiffUtil.ItemCallback<Contest>() {
        override fun areItemsTheSame(oldItem: Contest, newItem: Contest): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Contest, newItem: Contest): Boolean {
            return oldItem == newItem
        }

    }
}